provisioning/service/src/HttpDeviceRegistrationClient.cs

```diff
@@ -0,0 +1,660 @@
+        }
+
+        private static Uri GetQueryEnrollmentUri()
+        {
+            return new Uri(QueryEnrollmentUriFormat.FormatInvariant(ApiVersionQueryString), UriKind.Relative);
+        }
+
+        private static Uri GetQueryEnrollmentGroupUri()
+        {
+            return new Uri(QueryEnrollmentGroupUriFormat.FormatInvariant(ApiVersionQueryString), UriKind.Relative);
+        }
+
+        private static Uri GetEnrollmentGroupUri(string enrollmentGroupId)
+        {
+            enrollmentGroupId = WebUtility.UrlEncode(enrollmentGroupId);
+            return new Uri(EnrollmentGroupUriFormat.FormatInvariant(enrollmentGroupId, ApiVersionQueryString), UriKind.Relative);
+        }
+
+        private static Uri GetDeviceRegistrationUri(string registrationId)
+        {
+            registrationId = WebUtility.UrlEncode(registrationId);
+            return new Uri(DeviceRegistrationUriFormat.FormatInvariant(registrationId, ApiVersionQueryString), UriKind.Relative);
+        }
+
+        private static Uri GetDeviceRegistrationsUri(string enrollmentGroupId)
+        {
+            enrollmentGroupId = WebUtility.UrlEncode(enrollmentGroupId);
+            return new Uri(DeviceRegistrationsUriFormat.FormatInvariant(enrollmentGroupId, ApiVersionQueryString), UriKind.Relative);
+        }
+
+        static void ValidateRegistrationAndDeviceId(Enrollment enrollment)
+        {
+            if (enrollment == null)
+            {
+                throw new ArgumentNullException(nameof(enrollment));
+            }
+
+            if (string.IsNullOrWhiteSpace(enrollment.RegistrationId))
+            {
+                throw new ArgumentException("enrollment.RegistrationId");
+            }
+
+            if (!DeviceIdRegex.IsMatch(enrollment.RegistrationId))
+            {
+                throw new ArgumentException(ApiResources.EnrollmentIdNotValid.FormatInvariant(enrollment.RegistrationId));
+            }
+
+            if(!string.IsNullOrWhiteSpace(enrollment.DeviceId) &&
+               !DeviceIdRegex.IsMatch(enrollment.DeviceId))
+            {
+                throw new ArgumentException(ApiResources.EnrollmentIdNotValid.FormatInvariant(enrollment.DeviceId));
+            }
+        }
+
+        static void Validate(EnrollmentGroup enrollmentGroup)
+        {
+            if (enrollmentGroup == null)
+            {
+                throw new ArgumentNullException(nameof(enrollmentGroup));
+            }
+
+            if (string.IsNullOrWhiteSpace(enrollmentGroup.EnrollmentGroupId))
+            {
+                throw new ArgumentException("enrollmentGroup.EnrollmentGroupId");
+            }
+        }
+
+        async Task<DpsQueryResult> ExecuteQueryAsync(Uri queryUri, int? pageSize, string continuationToken, CancellationToken cancellationToken)
+        {
+            this.EnsureInstanceNotClosed();
+
+            //todo: support sqlQueryStrings
```

Comment:
Sounds like we should add a GH issue and reference it here for these kinds of TODOs so we don't forget we have them.

Reply:
created Task 1580139 to track

---
In reply to: [141726374](https://github.com/Azure/azure-iot-sdk-csharp/pull/174#discussion_r141726374) [](ancestors = 141726374)